File: lua/test/test_read.lua
local d = require("mempeep.descriptors")
local read = require("mempeep.read")
local memory = require("mempeep.test.memory")
local ok_tracer = require("mempeep.tracers.ok_tracer")

-- ---------------------------------------------------------------------------
-- Integers
-- ---------------------------------------------------------------------------

do
  local reader = memory.mock_memory_reader("I4", "\x44\x33\x22\x11")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Int32, 0, reader, tracer)
  assert(ok)
  assert(v == 0x11223344)
end

-- non-zero offset read
do
  local reader = memory.mock_memory_reader("I4", "\xFF\xFF\x44\x33\x22\x11\xFF\xFF")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Int32, 2, reader, tracer)
  assert(ok)
  assert(v == 0x11223344)
end

-- i8 (signed, -1)
do
  local reader = memory.mock_memory_reader("I4", "\xFF")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Int8, 0, reader, tracer)
  assert(ok)
  assert(v == -1)
end

-- i16
do
  local reader = memory.mock_memory_reader("I4", "\xE8\x03")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Int16, 0, reader, tracer)
  assert(ok)
  assert(v == 1000)
end

-- i64
do
  local reader = memory.mock_memory_reader("I4", "\x15\xCD\x5B\x07\x00\x00\x00\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Int64, 0, reader, tracer)
  assert(ok)
  assert(v == 123456789)
end

-- scalar with unreadable address returns error
do
  local reader = memory.mock_memory_reader("I4", "")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Int32, 0, reader, tracer)
  assert(not ok)
  assert(v == nil)
end

-- ---------------------------------------------------------------------------
-- Struct
-- ---------------------------------------------------------------------------

-- Point{x=10, y=20}
do
  local Point = d.Struct("Point", { d.Field(d.Int32, "x"), d.Field(d.Int32, "y") })
  local reader = memory.mock_memory_reader("I4", "\x0A\x00\x00\x00\x14\x00\x00\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(Point, 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(v.x == 10)
  assert(v.y == 20)
end

-- Struct with Skip: a=1 at offset 0, 4 skip bytes, b=2 at offset 8
do
  local Skipped = d.Struct("Skipped", { d.Field(d.Int32, "a"), d.Skip(4), d.Field(d.Int32, "b") })
  local reader = memory.mock_memory_reader("I4", "\x01\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(Skipped, 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(v.a == 1)
  assert(v.b == 2)
end

-- Struct with Seek: a=7 at offset 0, b=99 at offset 8
do
  local Sparse = d.Struct("Sparse", { d.Field(d.Int32, "a"), d.Seek(8), d.Field(d.Int32, "b") })
  local reader = memory.mock_memory_reader("I4", "\x07\x00\x00\x00\x00\x00\x00\x00\x63\x00\x00\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(Sparse, 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(v.a == 7)
  assert(v.b == 99)
end

-- ---------------------------------------------------------------------------
-- Array
-- ---------------------------------------------------------------------------

-- array of 3 x i32: 10, 20, 30
do
  local reader = memory.mock_memory_reader("I4", "\x0A\x00\x00\x00\x14\x00\x00\x00\x1E\x00\x00\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Array(d.Int32, 3), 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(#v == 3)
  assert(v[1] == 10)
  assert(v[2] == 20)
  assert(v[3] == 30)
end

-- array of 2 x Point{i16 x, i16 y}: (1,2), (3,4)
do
  local Point = d.Struct("Point", { d.Field(d.Int16, "x"), d.Field(d.Int16, "y") })
  local reader = memory.mock_memory_reader("I4", "\x01\x00\x02\x00\x03\x00\x04\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Array(Point, 2), 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(#v == 2)
  assert(v[1].x == 1)
  assert(v[1].y == 2)
  assert(v[2].x == 3)
  assert(v[2].y == 4)
end

-- ---------------------------------------------------------------------------
-- Vector
-- ---------------------------------------------------------------------------

-- vector of 3 x i32: begin=8, end=20; elements 100, 200, 300
do
  local reader = memory.mock_memory_reader(
    "I4",
    "\x08\x00\x00\x00" -- begin = 8
      .. "\x14\x00\x00\x00" -- end   = 20
      .. "\x64\x00\x00\x00" -- [0] = 100
      .. "\xC8\x00\x00\x00" -- [1] = 200
      .. "\x2C\x01\x00\x00" -- [2] = 300
  )
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Vector(d.Int32, 0x1000), 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(#v == 3)
  assert(v[1] == 100)
  assert(v[2] == 200)
  assert(v[3] == 300)
end

-- Empty vector (begin == end)
do
  local reader = memory.mock_memory_reader("I4", "\x08\x00\x00\x00" .. "\x08\x00\x00\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Vector(d.Int32, 0x1000), 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(#v == 0)
end

-- Vector begin > end: error
do
  local reader = memory.mock_memory_reader("I4", "\x10\x00\x00\x00" .. "\x08\x00\x00\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Vector(d.Int32, 0x1000), 0, reader, tracer)
  assert(not ok)
  assert(v == nil)
end

-- Unreadable vector pointers
do
  local reader = memory.mock_memory_reader("I4", "")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.Vector(d.Int32, 0x1000), 0, reader, tracer)
  assert(not ok)
  assert(v == nil)
end

-- ---------------------------------------------------------------------------
-- Circular list
-- ---------------------------------------------------------------------------

-- Three-node circular list: A -> B -> C -> A
-- Each node: value(i32) at +0, next(weak ptr) at +4 => 8 bytes/node
-- head ptr at offset 0; node A at 4, B at 12, C at 20
do
  local Node = d.Struct("Node", { d.Field(d.Int32, "value"), d.Field(d.RawAddr(), "next") })
  local reader = memory.mock_memory_reader(
    "I4",
    "\x04\x00\x00\x00" -- head ptr = 4
      .. "\x01\x00\x00\x00\x0C\x00\x00\x00" -- A: value=1, next=12
      .. "\x02\x00\x00\x00\x14\x00\x00\x00" -- B: value=2, next=20
      .. "\x03\x00\x00\x00\x04\x00\x00\x00" -- C: value=3, next=4
  )
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.CircularList(Node, "next", 0x1000), 0, reader, tracer)
  assert(ok)
  assert(v)
  assert(#v == 3)
  assert(v[1].value == 1)
  assert(v[2].value == 2)
  assert(v[3].value == 3)
end

-- Empty circular list (head pointer is null)
do
  local Node = d.Struct("Node", { d.Field(d.Int32, "value"), d.Field(d.RawAddr(), "next") })
  local reader = memory.mock_memory_reader("I4", "\x00\x00\x00\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.CircularList(Node, "next", 0x1000), 0, reader, tracer)
  assert(ok)
  assert(v)
  assert(#v == 0)
end

-- Unreadable head pointer
do
  local Node = d.Struct("Node", { d.Field(d.Int32, "value"), d.Field(d.RawAddr(), "next") })
  local reader = memory.mock_memory_reader("I4", "")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.CircularList(Node, "next", 0x1000), 0, reader, tracer)
  assert(not ok)
  assert(v == nil)
end

-- ---------------------------------------------------------------------------
-- Nested struct
-- ---------------------------------------------------------------------------

do
  local Inner = d.Struct("Inner", { d.Field(d.Int32, "a"), d.Field(d.Int32, "b") })
  local Outer = d.Struct("Outer", { d.Field(Inner, "inner"), d.Field(d.Int32, "c") })
  local reader = memory.mock_memory_reader(
    "I4",
    "\x0B\x00\x00\x00" -- a = 11
      .. "\x16\x00\x00\x00" -- b = 22
      .. "\x21\x00\x00\x00" -- c = 33
  )
  local tracer = ok_tracer.new()
  local v, ok = read.read(Outer, 0, reader, tracer)
  assert(ok)
  assert(v)
  assert(v.inner.a == 11)
  assert(v.inner.b == 22)
  assert(v.c == 33)
end

-- ---------------------------------------------------------------------------
-- PrimitiveArray
-- ---------------------------------------------------------------------------

-- 1D: three int16s
do
  local reader = memory.mock_memory_reader("I4", "\x01\x00\x02\x00\x03\x00")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.PrimitiveArray("i2", {3}), 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(#v == 3)
  assert(v[1] == 1)
  assert(v[2] == 2)
  assert(v[3] == 3)
end

-- 2D: PrimArray("I1", {2, 3}) => std::array<std::array<uint8_t,3>,2>
do
  local reader = memory.mock_memory_reader("I4", "\x01\x02\x03\x04\x05\x06")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.PrimitiveArray("I1", {2, 3}), 0, reader, tracer)
  assert(ok)
  assert(v ~= nil)
  assert(#v == 2)
  assert(#v[1] == 3)
  assert(#v[2] == 3)
  assert(v[1][1] == 1)
  assert(v[1][2] == 2)
  assert(v[1][3] == 3)
  assert(v[2][1] == 4)
  assert(v[2][2] == 5)
  assert(v[2][3] == 6)
end

-- Fails cleanly on short read
do
  local reader = memory.mock_memory_reader("I4", "\x01\x02")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.PrimitiveArray("I1", {3, 2}), 0, reader, tracer)
  assert(not ok)
end

-- ---------------------------------------------------------------------------
-- ZString
-- ---------------------------------------------------------------------------

-- Null terminator present before max_len
do
  local reader = memory.mock_memory_reader("I1", "\2\0hello\0world")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.ZString(11), 0, reader, tracer)
  assert(ok)
  assert(v == "hello")
end

-- No null terminator within max_len: returns all max_len bytes as-is
do
  local reader = memory.mock_memory_reader("I1", "\1abcdefg\0")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.ZString(4), 0, reader, tracer)
  assert(not ok)
  assert(v == "abcd")
end

-- Null terminator at position 0: empty string
do
  local reader = memory.mock_memory_reader("I1", "\1\0abc")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.ZString(4), 0, reader, tracer)
  assert(ok)
  assert(v == "")
end

-- Non-zero start offset
do
  local reader = memory.mock_memory_reader("I1", "\xFF\xFF\3hi\0")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.ZString(3), 2, reader, tracer)
  assert(ok)
  assert(v == "hi")
end

-- Unreadable address
do
  local reader = memory.mock_memory_reader("I4", "")
  local tracer = ok_tracer.new()
  local v, ok = read.read(d.ZString(4), 0, reader, tracer)
  assert(not ok)
  assert(v == nil)
end

-- ZString inside a Struct: cursor lands after the fixed-size slot
do
  local S = d.Struct("S", { d.Field(d.ZString(4), "name"), d.Field(d.Int32, "value") })
  local reader = memory.mock_memory_reader("I4", "\x08\0\0\0\x2A\x00\x00\x00hi\0\0")
  local tracer = ok_tracer.new()
  local v, ok = read.read(S, 0, reader, tracer)
  assert(ok)
  assert(v.name == "hi")
  assert(v.value == 42)
end

File: cpp/test/test_read.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest/doctest.h>

#include <array>
#include <mempeep/read.hpp>
#include <mempeep/test/memory.hpp>
#include <mempeep/tracers/log_tracer.hpp>
#include <mempeep/tracers/ok_tracer.hpp>
#include <optional>
#include <string_view>

#include "support/mock_game.hpp"

using namespace mempeep;

// truly empty data
// (can't use "" as it includes null terminator)
static constexpr std::array<uint8_t, 0> empty_data{};

TEST_CASE("successful read") {
  auto reader = test::MockMemoryReader<uint8_t>{test::game_data};
  test::Game game{};
  auto tracer = LogTracer{OnLogEntryPrint{std::cout}, LogLevel::VALUES};
  CHECK(read(RemoteValue<test::TGame, uint8_t>{4}, reader, tracer, game));
  SUBCASE("level") { CHECK_EQ(game.level, 17); }
  SUBCASE("player") {
    CHECK_EQ(game.player.health, 123);
    CHECK_EQ(game.player.pos.x, 11);
    CHECK_EQ(game.player.pos.y, 22);
    CHECK_EQ(game.player.target_ptr, 0);
    CHECK_EQ(game.player.shop_ptr, 2);
    CHECK_EQ(game.player.weapon_ptr, 6);
    CHECK_EQ(game.player.prev_pos.x, 88);
    CHECK_EQ(game.player.prev_pos.y, 99);
    CHECK_EQ(game.player.mana, 47);
    SUBCASE("tagged_pos") {
      REQUIRE(game.player.tagged_pos.has_value());
      CHECK_EQ(game.player.tagged_pos->x, 55);
      CHECK_EQ(game.player.tagged_pos->y, 66);
    }
    CHECK(!game.player.house_pos.has_value());
  }
  SUBCASE("hands") {
    CHECK_EQ(game.hands[0].x, 1);
    CHECK_EQ(game.hands[0].y, 2);
    CHECK_EQ(game.hands[1].x, 3);
    CHECK_EQ(game.hands[1].y, 4);
  }
  SUBCASE("pets") {
    REQUIRE_EQ(game.pets.size(), 3);
    CHECK_EQ(game.pets[0].x, 5);
    CHECK_EQ(game.pets[0].y, 6);
    CHECK_EQ(game.pets[1].x, 7);
    CHECK_EQ(game.pets[1].y, 8);
    CHECK_EQ(game.pets[2].x, 9);
    CHECK_EQ(game.pets[2].y, 10);
  }
  SUBCASE("caves") {
    REQUIRE_EQ(game.caves.size(), 4);
    CHECK_EQ(game.caves[0].id, 16);
    CHECK_EQ(game.caves[1].id, 18);
    CHECK_EQ(game.caves[2].id, 20);
    CHECK_EQ(game.caves[3].id, 22);
  }
}

TEST_CASE("failed read: complete failure") {
  auto reader = test::MockMemoryReader<uint8_t>{empty_data};
  test::Game game{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<test::TGame, uint8_t>{0}, reader, tracer, game));
}

TEST_CASE("failed read: invalid addresses") {
  static constexpr std::string_view data{test::game_data, 23};
  auto reader = test::MockMemoryReader<uint8_t>{data};
  test::Game game{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<test::TGame, uint8_t>{4}, reader, tracer, game));
  SUBCASE("level") { CHECK_EQ(game.level, 17); }
  SUBCASE("player") {
    CHECK_EQ(game.player.health, 123);
    CHECK_EQ(game.player.pos.x, 11);
    CHECK_EQ(game.player.pos.y, 22);
    CHECK_EQ(game.player.target_ptr, 0);
    CHECK_EQ(game.player.shop_ptr, 2);
    CHECK_EQ(game.player.weapon_ptr, 6);
    CHECK_EQ(game.player.prev_pos.x, 0);  // failed
    CHECK_EQ(game.player.prev_pos.y, 0);  // failed
    CHECK_EQ(game.player.mana, 47);
    SUBCASE("tagged_pos") {
      REQUIRE(game.player.tagged_pos.has_value());  // pointer was not null
      CHECK(game.player.tagged_pos->x == 0);        // failed
      CHECK(game.player.tagged_pos->y == 0);        // failed
    }
    CHECK(!game.player.house_pos.has_value());
  }
}

TEST_CASE("failed read: skip overflow") {
  struct Overflow {};

  using TOverflow = Struct<Overflow, Fields<Skip<0xff>, Skip<0xff>>>;
  auto reader = test::MockMemoryReader<uint8_t>{empty_data};
  Overflow overflow{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<TOverflow, uint8_t>{0}, reader, tracer, overflow));
}

TEST_CASE("failed read: null ref") {
  struct Obj {
    uint8_t item;
  };

  using TObj = Struct<Obj, Fields<Field<Ref<UInt8>, &Obj::item>>>;
  auto reader = test::MockMemoryReader<uint8_t>{"\x00"};
  Obj obj{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<TObj, uint8_t>{0}, reader, tracer, obj));
}

TEST_CASE("failed read: missing ref") {
  struct Obj {
    uint8_t item;
  };

  using TObj = Struct<Obj, Fields<Field<Ref<UInt8>, &Obj::item>>>;
  auto reader = test::MockMemoryReader<uint8_t>{empty_data};
  Obj obj{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<TObj, uint8_t>{0}, reader, tracer, obj));
}

TEST_CASE("failed read: missing nullable ref") {
  struct Obj {
    std::optional<uint8_t> item;
  };

  using TObj = Struct<Obj, Fields<Field<NullableRef<UInt8>, &Obj::item>>>;
  auto reader = test::MockMemoryReader<uint8_t>{empty_data};
  Obj obj{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<TObj, uint8_t>{0}, reader, tracer, obj));
}

TEST_CASE("ZString: null terminator before max_len") {
  auto reader = test::MockMemoryReader<uint8_t>{"hello\0world"};
  std::string out{};
  OkTracer tracer{};
  CHECK(read(RemoteValue<ZString<11>, uint8_t>{0}, reader, tracer, out));
  CHECK_EQ(out, "hello");
}

TEST_CASE("ZString: no null terminator") {
  auto reader = test::MockMemoryReader<uint8_t>{"abcdefg\0"};
  std::string out{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<ZString<4>, uint8_t>{0}, reader, tracer, out));
  CHECK_EQ(out, "abcd");
}

TEST_CASE("ZString: null at position 0") {
  auto reader = test::MockMemoryReader<uint8_t>{"\0abc"};
  std::string out{};
  OkTracer tracer{};
  CHECK(read(RemoteValue<ZString<4>, uint8_t>{0}, reader, tracer, out));
  CHECK_EQ(out, "");
}

TEST_CASE("ZString: unreadable address") {
  auto reader = test::MockMemoryReader<uint8_t>{empty_data};
  std::string out{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<ZString<4>, uint8_t>{0}, reader, tracer, out));
}

TEST_CASE("ZString: inside struct, cursor lands after string") {
  struct S {
    std::string name;
    int32_t value;
  };

  using TS
    = Struct<S, Fields<Field<ZString<4>, &S::name>, Field<Int32, &S::value>>>;
  auto reader = test::MockMemoryReader<uint32_t>{"hi\0\0\x2A\x00\x00\x00"};
  S s{};
  OkTracer tracer{};
  CHECK(read(RemoteValue<TS, uint32_t>{0}, reader, tracer, s));
  CHECK_EQ(s.name, "hi");
  CHECK_EQ(s.value, 42);
}

TEST_CASE("RemoteAddr: deferred read of an integer") {
  auto reader = test::MockMemoryReader<uint8_t>{
    "\x00\x00"
    "\x44\x33\x22\x11"
  };
  // read address only
  RemoteValue<Int32, uint8_t> out{};
  OkTracer tracer{};
  CHECK(read(
    RemoteValue<RemoteAddr<Int32, uint8_t>, uint8_t>{2}, reader, tracer, out
  ));
  CHECK_EQ(out.address, 2);
  // read full structure
  int32_t value{};
  OkTracer tracer2{};
  CHECK(read(out, reader, tracer2, value));
  CHECK_EQ(value, 0x11223344);
}

TEST_CASE("RemoteAddr: deferred read of ZString") {
  auto reader = test::MockMemoryReader<uint8_t>{"hello\0world"};
  // read address only
  RemoteValue<ZString<6>, uint8_t> out{};
  OkTracer tracer{};
  CHECK(read(
    RemoteValue<RemoteAddr<ZString<6>, uint8_t>, uint8_t>{0},
    reader,
    tracer,
    out
  ));
  CHECK_EQ(out.address, 0);
  // read full structure
  std::string value{};
  OkTracer tracer2{};
  CHECK(read(out, reader, tracer2, value));
  CHECK_EQ(value, "hello");
}

TEST_CASE("RemoteAddr: inside struct, cursor advances past descriptor bytes") {
  struct S {
    RemoteValue<Int32, uint8_t> data;
    int16_t after;
  };
  using TS = Struct<
    S,
    Fields<
      Field<RemoteAddr<Int32, uint8_t>, &S::data>,
      Field<Int16, &S::after>>>;
  auto reader = test::MockMemoryReader<uint8_t>{"\x44\x33\x22\x11\x77\x66"};
  // read address only
  S s{};
  OkTracer tracer{};
  CHECK(read(RemoteValue<TS, uint8_t>{0}, reader, tracer, s));
  CHECK_EQ(s.data.address, 0);
  CHECK_EQ(s.after, 0x6677);
  // read data
  int32_t data{};
  OkTracer tracer2{};
  CHECK(read(s.data, reader, tracer2, data));
  CHECK_EQ(data, 0x11223344);
}

TEST_CASE("Negative seek") {
  struct Obj {
    int16_t a;
    int16_t b;
  };

  using TObj = Struct<
    Obj,
    Fields<Seek<-2>, Field<Int16, &Obj::a>, Field<Int16, &Obj::b>>>;
  static_assert(byte_size<TObj, uint32_t>() == 2);
  auto reader = test::MockMemoryReader<uint32_t>{"\x11\x00\x22\x00"};
  Obj obj{};
  OkTracer tracer{};
  CHECK(read(RemoteValue<TObj, uint32_t>{2}, reader, tracer, obj));
  CHECK_EQ(obj.a, 0x11);
  CHECK_EQ(obj.b, 0x22);
}

TEST_CASE("Negative skip") {
  struct Obj {
    int32_t a;
    int32_t b;
  };
  using TObj = Struct<
    Obj,
    Fields<Field<Int32, &Obj::a>, Skip<-2>, Field<Int32, &Obj::b>>>;
  static_assert(byte_size<TObj, uint32_t>() == 6);
  auto reader = test::MockMemoryReader<uint32_t>{"\x01\x00\x02\x00\x00\x00"};
  Obj obj{};
  OkTracer tracer{};
  CHECK(read(RemoteValue<TObj, uint32_t>{0}, reader, tracer, obj));
  CHECK_EQ(obj.a, 0x020001);
  CHECK_EQ(obj.b, 0x02);
}

TEST_CASE("Seek underflow") {
  struct Obj {
    int32_t a;
  };
  using TObj = Struct<Obj, Fields<Seek<-1>, Field<Int32, &Obj::a>>>;
  static_assert(byte_size<TObj, uint32_t>() == 3);
  auto reader = test::MockMemoryReader<uint32_t>{"\x01\x00\x00\x00"};
  Obj obj{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<TObj, uint32_t>{0}, reader, tracer, obj));
}

TEST_CASE("Skip underflow") {
  struct Obj {
    int32_t a;
  };
  using TObj = Struct<Obj, Fields<Field<Int32, &Obj::a>, Skip<-8>, Seek<4>>>;
  static_assert(byte_size<TObj, uint32_t>() == 4);
  auto reader = test::MockMemoryReader<uint32_t>{"\x01\x00\x00\x00"};
  Obj obj{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<TObj, uint32_t>{0}, reader, tracer, obj));
}

TEST_CASE("Skip overflow") {
  struct Obj {
    int32_t a;
  };
  using TObj = Struct<Obj, Fields<Field<Int32, &Obj::a>, Skip<256>>>;
  static_assert(byte_size<TObj, uint8_t>() == 260);
  auto reader = test::MockMemoryReader<uint8_t>{"\x01\x00\x00\x00"};
  Obj obj{};
  OkTracer tracer{};
  CHECK(!read(RemoteValue<TObj, uint8_t>{0}, reader, tracer, obj));
}
